compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -473,6 +473,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let mut custom_span_label = false;
 
         let static_candidates = &mut no_match_data.static_candidates;
+
+        static_candidates.dedup();
```

Comment:
Could you add a comment why we cannot remove the duplicated call to `record_static_candidate`?

Reply:
`inherent` and `extension` may append the same item to `static_candidates`.

For example:

```rs
trait Cat {
    fn nya() {}
}

fn uwu<T: Cat>(c: T) {
    c.nya();
}
```

Both `<T as Cat>` (inherent) and `<_ as Cat>` (extension) had been tried.

And another:

```rs
trait Cat {
    fn nya() {}
}

fn uwu<T>(c: T) {
    c.nya();
}
```

`<_ as Cat>` (extension) should be tried.

Therefore, we can't remove duplicated call of `record_static_candidate`.

Reply:
The comment you added says that in order to improve the diagnostic, we record both inherent and extension cases. But this PR does the opposite: we deduplicate to improve the diagnostic.
You don't need to link to the PR. The 2 lines-long explanation is enough.

Reply:
I have two questions:
1. Is it correct to say that we cannot remove the duplicated call of ﻿`record_static_candidate`? Also, is it intended behavior that the unexpected diagnostic of #103646 was caused by invoking ﻿`record_static_candidate` for the same item in both ﻿`assemble_extension_candidates_for_trait` and ﻿`assemble_inherent_impl_probe`?
2. Is ﻿`static_candidates.dedup()` the right approach here? If this statement has too broad impact, can we change the condition to ﻿`!self.has_applicable_self(&item) && !self.static_candidates.contains(&item)` in [extension_candidates](https://github.com/rust-lang/rust/blob/master/compiler/rustc_hir_typeck/src/method/probe.rs#L968)?